zdm/grid.py

```diff
@@ -0,0 +1,665 @@
+        iE2 = iE1+1
+        # iE1 should never be the highest energy, since it will always have a probability of 0
+        kE1=(r-PEs[iE2])/(PEs[iE1]-PEs[iE2])
+        kE2=1.-kE1
+        MCE=10**(np.log10(Es[iE1])*kE1 + np.log10(Es[iE2])*kE2)
+        MCs=MCE/Eth
+        
+        FRBparams=[MCz,MCDM,MCb,j,MCs]
+        return FRBparams,pwb
+        
+
+    def update(self, vparams:dict, ALL=False):
+        """Update the grid based on a set of input
+        parameters
+        
+        Hierarchy:
+        Each indent corresponds to one 'level'.
+        This is used in the program control below
+        to dictate how far each tree should proceed
+        in calculation.
+        Direct variable inputs are always listed first
+        We see that sfr evolution and dm smearing
+        lie just before the pdv step
+        Hence, we deal with these first, and
+        calc rates as a final step regardless
+        of what else has changed.
+
+        Args:
+            vparams (dict):  dict containing the parameters
+                to be updated and their values
+            ALL (bool, optional):  If True, update the full grid
+        
+        calc_rates:
+            calc_pdv
+                Emin
+                Emax
+                gamma
+                calc_thresholds
+                    F0
+                    alpha
+                    bandwidth
+            set_evolution
+                sfr_n
+            
+            smear_grid
+                grid
+                mask
+                    dmx_params (lmean, lsigma)
+            dV
+            zdm_grid
+                H0
+        
+        Note that the grid is independent of the constant C (trivial dependence)
+
+        Args:
+            vparams (dict): [description]
+        """
+        # Init
+        reset_cos, get_zdm, calc_dV = False, False, False
+        smear_mask, smear_dm, calc_pdv, set_evol = False, False, False, False
+        new_sfr_smear, new_pdv_smear, calc_thresh = False, False, False
+
+        # Cosmology -- Only H0 so far
+        if self.chk_upd_param('H0', vparams, update=True):
+            reset_cos = True
+            get_zdm = True
+            calc_dV = True
+            smear_mask = True
+            smear_dm = True
+            calc_thresh = True
+            calc_pdv = True
+            set_evol = True
+            new_sfr_smear = True
+
+        # Mask?
+        # IT IS IMPORTANT TO USE np.any so that each item is executed!!
+        if np.any([self.chk_upd_param('lmean', vparams, update=True), 
+            self.chk_upd_param('lsigma', vparams, update=True)]):
+            smear_mask = True
+            smear_dm = True
+            new_sfr_smear=True
+
+
+        # SFR?
+        if self.chk_upd_param('sfr_n', vparams, update=True):
+            set_evol = True
+            new_sfr_smear=True  # True for either alpha_method
+        if self.chk_upd_param('alpha', vparams, update=True):
```

Comment:
if alpha_method==0, we need
calc_pdv=True
new_pdv_smear=True
calc_thresh=True [since thresholds change]
if alpha_method==1,  we only need to set
set_evol=True
new_sfr_smear=True [since effectively this is changing source evolution]

Reply:
got it and this is now matching what was
in `update_grid`